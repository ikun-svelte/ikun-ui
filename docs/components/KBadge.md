---
title: KBadge
lang: en-US
---

# KBadge

A number or status mark on buttons and icons.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/badge
```

```bash [yarn]
yarn add @ikun-ui/badge
```

```bash [npm]
npm install @ikun-ui/badge
```

:::

## Basic usage

Displays the amount of new messages.

The amount is defined with value which accepts Number or String.

<demo src="badge/basic.svelte"  github='Badge'></demo>

## Max Value

You can customize the max value.

The max value is defined by property max which is a Number. Note that it only works when value is also a Number.

<demo src="badge/max-value.svelte" github='Badge'></demo>

## Customizations

Displays text content other than numbers.

When value is a String, it can display customized text.

<demo src="badge/customizations.svelte" github='Badge'></demo>

## Dot Mark

Use a dot to mark content that needs to be noticed.

Use the attribute isDot. It is a Boolean.

<demo src="badge/dot-mark.svelte" github='Badge'></demo>

## Badge Props

| Name  | Type                                                 | Default | Description                                                                 |
| ----- | ---------------------------------------------------- | ------- | --------------------------------------------------------------------------- |
| value | `string` / `number`                                  | `''`    | display value.                                                              |
| max   | `number`                                             | `99`    | maximum value, shows {max}+ when exceeded. Only works if value is a number. |
| isDot | `boolean`                                            | `false` | if a little dot is displayed.                                               |
| show  | `boolean`                                            | `true`  | show badge.                                                                 |
| type  | `primary` / `success` / `error` / `warning` / `info` | `error` | badge type.                                                                 |
| cls   | `string`                                             | `''`    | Additional class for component                                              |
| attrs | `Record<string, string>`                             | `{}`    | Additional attributes                                                       |

## Badge Slots

| Name    | Description               |
| ------- | ------------------------- |
| default | Customize default content |
