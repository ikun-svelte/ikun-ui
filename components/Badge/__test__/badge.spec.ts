import { afterEach, expect, test, describe, beforeEach } from 'vitest';
import KBadge from '../src';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KBadge', () => {
	test('props: cls', async () => {
		const instance = new KBadge({
			target: host,
			props: {
				cls: 'k-badge--test'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-badge--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: value is number', async () => {
		const value = 23;
		const instance = new KBadge({
			target: host,
			props: {
				value
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('sup')?.textContent).toBe(`${value}`);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: value is string', async () => {
		const value = 'new';
		const instance = new KBadge({
			target: host,
			props: {
				value
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('sup')?.textContent).toBe(value);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: max', async () => {
		const value = 102;
		const max = 99;
		const instance = new KBadge({
			target: host,
			props: {
				value,
				max
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('sup')?.textContent).toBe(`${max}+`);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: isDot', async () => {
		const isDot = true;
		const instance = new KBadge({
			target: host,
			props: {
				isDot
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-badge--content-dot')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: show', async () => {
		const show = false;
		const instance = new KBadge({
			target: host,
			props: {
				show
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('sup')).toBeNull();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: type', async () => {
		const types = ['primary', 'success', 'error', 'warning', 'info'];
		const value = 102;
		for (const type of types) {
			host.remove();
			initHost();
			const instance = new KBadge({
				target: host,
				props: {
					type,
					value
				}
			});
			expect(instance).toBeTruthy();
			expect((host as HTMLElement)!.innerHTML.includes(`k-badge--content-${type}`)).toBeTruthy();
			expect(host.innerHTML).matchSnapshot();
		}
	});
});
