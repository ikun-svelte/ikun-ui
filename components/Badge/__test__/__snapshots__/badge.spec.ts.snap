// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KBadge > props: cls 1`] = `"<div class=\\"k-badge k-badge--base k-badge--test\\"> </div>"`;

exports[`Test: KBadge > props: isDot 1`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-error k-badge--content-fixed k-badge--content-dot k-badge--content__dark\\"></sup></div>"`;

exports[`Test: KBadge > props: max 1`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-error k-badge--content-fixed k-badge--content__dark\\">99+</sup></div>"`;

exports[`Test: KBadge > props: show 1`] = `"<div class=\\"k-badge k-badge--base\\"> </div>"`;

exports[`Test: KBadge > props: type 1`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-primary k-badge--content-fixed k-badge--content__dark\\">99+</sup></div>"`;

exports[`Test: KBadge > props: type 2`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-success k-badge--content-fixed k-badge--content__dark\\">99+</sup></div>"`;

exports[`Test: KBadge > props: type 3`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-error k-badge--content-fixed k-badge--content__dark\\">99+</sup></div>"`;

exports[`Test: KBadge > props: type 4`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-warning k-badge--content-fixed k-badge--content__dark\\">99+</sup></div>"`;

exports[`Test: KBadge > props: type 5`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-info k-badge--content-fixed k-badge--content__dark\\">99+</sup></div>"`;

exports[`Test: KBadge > props: value is number 1`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-error k-badge--content-fixed k-badge--content__dark\\">23</sup></div>"`;

exports[`Test: KBadge > props: value is string 1`] = `"<div class=\\"k-badge k-badge--base\\"> <sup class=\\"k-badge--content k-badge--content-error k-badge--content-fixed k-badge--content__dark\\">new</sup></div>"`;
